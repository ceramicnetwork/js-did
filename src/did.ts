import { ResolverOptions, DIDResolutionResult, ResolverRegistry, Resolver } from 'did-resolver'
import { createJWE, JWE, verifyJWS, resolveX25519Encrypters } from 'did-jwt'
import { encodePayload, prepareCleartext, decodeCleartext } from 'dag-jose-utils'
import { RPCClient } from 'rpc-utils'
import { CID } from 'multiformats/cid'
import { CacaoBlock, Cacao } from 'ceramic-cacao'
import type { DagJWS, DIDProvider, DIDProviderClient } from './types.js'
import {
  fromDagJWS,
  encodeBase64,
  base64urlToJSON,
  decodeBase64,
  encodeBase64Url,
  randomString,
  didWithTime,
  extractControllers,
} from './utils.js'

export interface AuthenticateOptions {
  provider?: DIDProvider
  aud?: string
  paths?: Array<string>
}

export interface AuthenticateParams {
  nonce: string
  aud?: string
  paths?: Array<string>
}

export interface AuthenticateResponse extends AuthenticateParams {
  did: string
  exp: number
}

export interface CreateJWSOptions {
  did?: string
  protected?: Record<string, any>
  linkedBlock?: string
}

export interface VerifyJWSOptions {
  /**
   * JS timestamp when the signature was allegedly made. `undefined` means _now_.
   */
  atTime?: Date

  /**
   * If true, timestamp checking is disabled.
   */
  disableTimecheck?: boolean

  /**
   * DID that issued the signature.
   */
  issuer?: string

  /**
   * Cacao OCAP to verify the JWS with.
   */
  capability?: Cacao

  /**
   * Number of seconds that a revoked key stays valid for after it was revoked
   */
  revocationPhaseOutSecs?: number
}

export interface VerifyJWSResult {
  kid: string
  payload?: Record<string, any>
  didResolutionResult: DIDResolutionResult
}

export interface CreateJWEOptions {
  protectedHeader?: Record<string, any>
  aad?: Uint8Array
}

export interface DecryptJWEOptions {
  did?: string
}

export interface DecryptJWEResult {
  cleartext: string // base64-encoded
}

export interface DagJWSResult {
  jws: DagJWS
  linkedBlock: Uint8Array
  cacaoBlock?: Uint8Array
}

export interface DIDOptions {
  provider?: DIDProvider
  resolver?: Resolver | ResolverRegistry
  resolverOptions?: ResolverOptions
  capability?: Cacao
}

function isResolver(resolver: Resolver | ResolverRegistry): resolver is Resolver {
  return 'registry' in resolver && 'cache' in resolver
}

/**
 * Interact with DIDs.
 */
export class DID {
  private _client?: DIDProviderClient
  private _id?: string
  private _resolver!: Resolver
  private _capability?: Cacao

  constructor({ provider, resolver = {}, resolverOptions, capability }: DIDOptions = {}) {
    if (provider != null) {
      this._client = new RPCClient(provider)
    }
    if (capability) {
      this._capability = capability
    }
    this.setResolver(resolver, resolverOptions)
  }

  /**
   * Check if the DID has a capability attached
   */
  get capability(): Cacao {
    if (!this._capability) {
      throw new Error('DID has no capability attached')
    }
    return this._capability
  }

  /**
   * Check if user is authenticated.
   */
  get authenticated(): boolean {
    return this._id != null
  }

  /**
   * Get the DID identifier of the user.
   */
  get id(): string {
    if (this._id == null) {
      throw new Error('DID is not authenticated')
    }
    return this._id
  }

  /**
   * Attach a capability to the DID instance
   * @param cap The capability to attach
   * @returns A new DID instance with the capability attached
   */
  withCapability(cap: Cacao): DID {
    return new DID({
      provider: this._client?.connection,
      resolver: this._resolver,
      capability: cap,
    })
  }

  /**
   * Set the DID provider of this instance.
   * Only callable if provider not already set.
   *
   * @param provider    The DIDProvider to use
   */
  setProvider(provider: DIDProvider): void {
    if (this._client == null) {
      this._client = new RPCClient(provider)
    } else if (this._client.connection !== provider) {
      throw new Error(
        'A different provider is already set, create a new DID instance to use another provider'
      )
    }
  }

  /**
   * Set the DID-resolver user by this instance
   *
   * @param resolver            Either a Resolver instance or an object with specific resolvers
   * @param resolverOptions     Options to use for the created resolver. Will be ignored if a Resolver instance is passed
   */
  setResolver(resolver: Resolver | ResolverRegistry, resolverOptions?: ResolverOptions): void {
    this._resolver = isResolver(resolver) ? resolver : new Resolver(resolver, resolverOptions)
  }

  /**
   * Authenticate the user.
   */
  async authenticate({ provider, paths = [], aud }: AuthenticateOptions = {}): Promise<string> {
    if (provider != null) {
      this.setProvider(provider)
    }
    if (this._client == null) {
      throw new Error('No provider available')
    }
    const nonce = randomString()
    const jws = await this._client.request('did_authenticate', {
      nonce,
      aud,
      paths,
    })
    const { kid } = await this.verifyJWS(jws)
    const payload = base64urlToJSON(jws.payload) as AuthenticateResponse
    if (!kid.includes(payload.did)) throw new Error('Invalid authencation response, kid mismatch')
    if (payload.nonce !== nonce) throw new Error('Invalid authencation response, wrong nonce')
    if (payload.aud !== aud) throw new Error('Invalid authencation response, wrong aud')
    if (payload.exp < Date.now() / 1000) throw new Error('Invalid authencation response, expired')
    this._id = payload.did
    return this._id
  }

  /**
   * Create a JWS encoded signature over the given payload.
   * Will be signed by the currently authenticated DID.
   *
   * @param payload             The payload to sign
   * @param options             Optional parameters
   */
  async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
    if (this._client == null) throw new Error('No provider available')
    if (this._id == null) throw new Error('DID is not authenticated')
    if (this._capability) {
      const cacaoBlock = await CacaoBlock.fromCacao(this._capability)
      const capCID = CID.asCID(cacaoBlock.cid)
      if (!capCID) {
        throw new Error(
          `Capability CID of the JWS cannot be set to the capability payload cid as they are incompatible`
        )
      }
      options.protected = options.protected || {}
      options.protected.cap = `ipfs://${capCID?.toString()}`
    }
    const { jws } = await this._client.request('did_createJWS', {
      did: this._id,
      ...options,
      payload,
    })
    return jws
  }

  /**
   * Create an IPFS compatibe DagJWS encoded signature over the given payload.
   * Will be signed by the currently authenticated DID.
   *
   * @param payload             The payload to sign, may include ipld links
   * @param options             Optional parameters
   */
  async createDagJWS(
    payload: Record<string, any>,
    options: CreateJWSOptions = {}
  ): Promise<DagJWSResult> {
    const { cid, linkedBlock } = await encodePayload(payload)
    const payloadCid = encodeBase64Url(cid.bytes)
    Object.assign(options, { linkedBlock: encodeBase64(linkedBlock) })
    const jws = await this.createJWS(payloadCid, options)

    const compatibleCID = CID.asCID(cid)
    if (!compatibleCID) {
      throw new Error(
        'CID of the JWS cannot be set to the encoded payload cid as they are incompatible'
      )
    }
    jws.link = compatibleCID

    if (this._capability) {
      const cacaoBlock = await CacaoBlock.fromCacao(this._capability)
      return { jws, linkedBlock, cacaoBlock: cacaoBlock.bytes }
    }
    return { jws, linkedBlock }
  }

  /**
   * Verify a JWS. Uses the 'kid' in the header as the way to resolve
   * the author public key.
   *
   * @param jws                 The JWS to verify
   * @param options             Optional arguments for verification
   * @returns                   Information about the signed JWS
   */
  async verifyJWS(jws: string | DagJWS, options: VerifyJWSOptions = {}): Promise<VerifyJWSResult> {
    if (typeof jws !== 'string') jws = fromDagJWS(jws)
    const kid = base64urlToJSON(jws.split('.')[0]).kid as string
    if (!kid) throw new Error('No "kid" found in jws')
    const didResolutionResult = await this.resolve(kid)
    const timecheckEnabled = !options.disableTimecheck
    if (timecheckEnabled) {
      const nextUpdate = didResolutionResult.didDocumentMetadata?.nextUpdate
      if (nextUpdate) {
        // This version of the DID document has been revoked. Check if the JWS
        // was signed before the revocation happened.
        const phaseOutMS = options.revocationPhaseOutSecs
          ? options.revocationPhaseOutSecs * 1000
          : 0
        const revocationTime = new Date(nextUpdate).valueOf() + phaseOutMS
        const isEarlier = options.atTime && options.atTime.getTime() < revocationTime
        const isLater = !isEarlier
        if (isLater) {
          // Do not allow using a key _after_ it is being revoked
          throw new Error(`invalid_jws: signature authored with a revoked DID version: ${kid}`)
        }
      }
      // Key used before `updated` date
      const updated = didResolutionResult.didDocumentMetadata?.updated
      if (updated && options.atTime && options.atTime.getTime() < new Date(updated).valueOf()) {
        throw new Error(`invalid_jws: signature authored before creation of DID version: ${kid}`)
      }
    }

    const signerDid = didResolutionResult.didDocument?.id
    if (
      options.issuer &&
      options.issuer === options.capability?.p.iss &&
      signerDid === options.capability.p.aud
    ) {
      Cacao.verify(options.capability, {
        atTime: options.atTime ? options.atTime : undefined,
      })
    } else if (options.issuer && options.issuer !== signerDid) {
      const issuerUrl = didWithTime(options.issuer, options.atTime)
      const issuerResolution = await this.resolve(issuerUrl)
      const controllerProperty = issuerResolution.didDocument?.controller
      const controllers = extractControllers(controllerProperty)

      if (
        options.capability?.s &&
        options.capability.p.aud === signerDid &&
        controllers.includes(options.capability.p.iss)
      ) {
        Cacao.verify(options.capability, {
          atTime: options.atTime ? options.atTime : undefined,
        })
      } else {
        const signerIsController = signerDid ? controllers.includes(signerDid) : false
        if (!signerIsController) {
          throw new Error(`invalid_jws: not a valid verificationMethod for issuer: ${kid}`)
        }
      }
    }

    const publicKeys = didResolutionResult.didDocument?.verificationMethod || []
    // verifyJWS will throw an error if the signature is invalid
    verifyJWS(jws, publicKeys)
    let payload
    try {
      payload = base64urlToJSON(jws.split('.')[1])
    } catch (e) {
      // If an error is thrown it means that the payload is a CID.
    }
    return { kid, payload, didResolutionResult }
  }

  /**
   * Create a JWE encrypted to the given recipients.
   *
   * @param cleartext           The cleartext to be encrypted
   * @param recipients          An array of DIDs
   * @param options             Optional parameters
   */
  async createJWE(
    cleartext: Uint8Array,
    recipients: Array<string>,
    options: CreateJWEOptions = {}
  ): Promise<JWE> {
    const encrypters = await resolveX25519Encrypters(recipients, this._resolver)
    return createJWE(cleartext, encrypters, options.protectedHeader, options.aad)
  }

  /**
   * Create an IPFS compatibe DagJWE encrypted to the given recipients.
   *
   * @param cleartext           The cleartext to be encrypted, may include ipld links
   * @param recipients          An array of DIDs
   * @param options             Optional parameters
   */
  async createDagJWE(
    cleartext: Record<string, any>,
    recipients: Array<string>,
    options: CreateJWEOptions = {}
  ): Promise<JWE> {
    const preparedCleartext = await prepareCleartext(cleartext)
    return this.createJWE(preparedCleartext, recipients, options)
  }

  /**
   * Try to decrypt the given JWE with the currently authenticated user.
   *
   * @param jwe                 The JWE to decrypt
   * @param options             Optional parameters
   */
  async decryptJWE(jwe: JWE, options: DecryptJWEOptions = {}): Promise<Uint8Array> {
    if (this._client == null) throw new Error('No provider available')
    if (this._id == null) throw new Error('DID is not authenticated')
    const { cleartext } = await this._client.request('did_decryptJWE', {
      did: this._id,
      ...options,
      jwe,
    })
    return decodeBase64(cleartext)
  }

  /**
   * Try to decrypt the given DagJWE with the currently authenticated user.
   *
   * @param jwe                 The JWE to decrypt
   * @returns                   An ipld object
   */
  async decryptDagJWE(jwe: JWE): Promise<Record<string, any>> {
    const bytes = await this.decryptJWE(jwe)
    return decodeCleartext(bytes)
  }

  /**
   * Resolve the DID Document of the given DID.
   *
   * @param didUrl              The DID to resolve
   */
  async resolve(didUrl: string): Promise<DIDResolutionResult> {
    const result = await this._resolver.resolve(didUrl)
    if (result.didResolutionMetadata.error) {
      const { error, message } = result.didResolutionMetadata
      const maybeMessage = message ? `, ${message as string}` : ''
      throw new Error(`Failed to resolve ${didUrl}: ${error}${maybeMessage}`)
    }
    return result
  }
}
