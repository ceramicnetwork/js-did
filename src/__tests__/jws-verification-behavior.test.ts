import { DID } from '../did'

// Using authSecret: `sha256.hash(uint8arrays.fromString(`HELLOWORLD`));

const VERSION_0_VANILLA = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#AdUB2e27TjdFwth',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '6wbGfRTmgTcnjTEHq6Wdg3AX2W8NKZTKJ3vLd166Ya7w',
      },
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#vwgkzudRSEdS2dW',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '2BQp6zUiYQCEFP7Aaph6vm9vYNXAndNu9q59q86ZWDYmn',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#vwgkzudRSEdS2dW',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '2BQp6zUiYQCEFP7Aaph6vm9vYNXAndNu9q59q86ZWDYmn',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#AdUB2e27TjdFwth',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '6wbGfRTmgTcnjTEHq6Wdg3AX2W8NKZTKJ3vLd166Ya7w',
      },
    ],
  },
  didDocumentMetadata: {
    versionId: '0',
  },
}

const VERSION_0_ROTATED = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#AdUB2e27TjdFwth',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '6wbGfRTmgTcnjTEHq6Wdg3AX2W8NKZTKJ3vLd166Ya7w',
      },
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#vwgkzudRSEdS2dW',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '2BQp6zUiYQCEFP7Aaph6vm9vYNXAndNu9q59q86ZWDYmn',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#vwgkzudRSEdS2dW',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '2BQp6zUiYQCEFP7Aaph6vm9vYNXAndNu9q59q86ZWDYmn',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#AdUB2e27TjdFwth',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '6wbGfRTmgTcnjTEHq6Wdg3AX2W8NKZTKJ3vLd166Ya7w',
      },
    ],
  },
  didDocumentMetadata: {
    nextUpdate: '2021-07-07T07:32:28Z',
    versionId: '0',
    nextVersionId: 'bafyreicveiixaitvrpocrgljdjw7ed3syp7b6evxcz4yzwwqaxnbn7tq2a',
  },
}

const VERSION_NEXT = {
  didResolutionMetadata: {
    contentType: 'application/did+json',
  },
  didDocument: {
    id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
    verificationMethod: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#Yu1U6tjUoVTM9tX',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: 'oJn8JX1LJKBmo4xy8bwRJjtAfbqm6zrUYB8wBTpL8g2o',
      },
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#MuzBPi9UD4cdq5q',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '45ii3puWe41GYUVbAw2KQjxAwcrRokjqJQzTykR5vTK5',
      },
    ],
    authentication: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#Yu1U6tjUoVTM9tX',
        type: 'EcdsaSecp256k1Signature2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: 'oJn8JX1LJKBmo4xy8bwRJjtAfbqm6zrUYB8wBTpL8g2o',
      },
    ],
    keyAgreement: [
      {
        id: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr#MuzBPi9UD4cdq5q',
        type: 'X25519KeyAgreementKey2019',
        controller: 'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr',
        publicKeyBase58: '45ii3puWe41GYUVbAw2KQjxAwcrRokjqJQzTykR5vTK5',
      },
    ],
  },
  didDocumentMetadata: {
    updated: '2021-07-07T07:32:28Z',
    versionId: 'bafyreicveiixaitvrpocrgljdjw7ed3syp7b6evxcz4yzwwqaxnbn7tq2a',
  },
}

// using: kid = did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr?version-id=0#vwgkzudRSEdS2dW
const jws =
  'eyJraWQiOiJkaWQ6MzpranpsNmN3ZTFqdzE0Nnl1YXBhOGNpemlqcmM0OWphbmF0ZjFnNmRuczVyenhlZjA5enFlZTd6enpxOTBmdHI_dmVyc2lvbi1pZD0wI3Z3Z2t6dWRSU0VkUzJkVyIsImFsZyI6IkVTMjU2SyJ9.eyJoZWxsbyI6IndvcmxkIn0.wLQ9QTfXRJe4stVzgOb6rXh6l913vycklrWoppcUkI888-B4o7jqlSxsQDBpEfDLqSr52A8wjV9TQmKqZ_qOHw'

test('vanilla version 0', async () => {
  const did = new DID()
  did.resolve = () => Promise.resolve(VERSION_0_VANILLA)
  const { kid } = await did.verifyJWS(jws)
  expect(kid).toEqual(
    'did:3:kjzl6cwe1jw146yuapa8cizijrc49janatf1g6dns5rzxef09zqee7zzzq90ftr?version-id=0#vwgkzudRSEdS2dW'
  )
})

test('rotated key', async () => {
  const did = new DID()
  did.resolve = (didUrl) => {
    const isVersion0 = /version-id=0/.exec(didUrl)
    if (isVersion0) {
      return Promise.resolve(VERSION_0_ROTATED)
    } else {
      return Promise.resolve(VERSION_NEXT)
    }
  }
  await expect(did.verifyJWS(jws)).rejects.toThrow(/JWS was signed with a revoked DID version/)
})

describe('atTime', () => {
  const did = new DID()
  const fauxResolve = (didUrl: string) => {
    const isVersion0 = /version-id=0/.exec(didUrl)
    if (isVersion0) {
      return Promise.resolve(VERSION_0_ROTATED)
    } else {
      return Promise.resolve(VERSION_NEXT)
    }
  }
  beforeEach(() => {
    did.resolve = fauxResolve
  })

  test('ok before rotation', async () => {
    const beforeRotation = new Date('2021-06-21T08:20:46Z').valueOf()
    const { kid } = await did.verifyJWS(jws, { atTime: beforeRotation })
    expect(kid).toMatchSnapshot()
  })
  test('fail after rotation', async () => {
    const afterRotation = new Date('2021-07-07T07:40:28Z').valueOf()
    await expect(did.verifyJWS(jws, { atTime: afterRotation })).rejects.toThrow(
      /JWS was signed with a revoked DID version/
    )
  })

  test('before DID version available', async () => {
    did.resolve = () => Promise.resolve(VERSION_NEXT)
    const beforeCreated = new Date('2021-07-01T00:00:00Z').valueOf()
    // Have lost keys for the old DID, so here is a new one.
    const jws =
      'eyJraWQiOiJkaWQ6MzpranpsNmN3ZTFqdzE0OXJ1cnFuaTRyMDY0YWh0MmFwbmJpZnJjMjlnaGZxam05bndkcGl6Nml4cmNuNnI0YXA_dmVyc2lvbi1pZD1iYWZ5cmVpZnRuNDJ6aHI3a2ZtdW9jYmthZXRlY2JrcG03anUyeHJlYXJlZmNsMnVhbGpmZWNnYjIycSNMV2RRZTVKQ0RGcGF5MUIiLCJhbGciOiJFUzI1NksifQ.eyJoZWxsbyI6IndvcmxkIn0.zXK7AKLADudA5UPBi4rgo1X3ZhMRT3wuWjZDfIwQ2VmB-Q7ZczXk8xJI-Wgv0K_YdDYmi5KslO9ObKCbe0fBmw'
    await expect(did.verifyJWS(jws, { atTime: beforeCreated })).rejects.toThrow(/not-yet created/)
  })
  test('before DID available for v0', async () => {
    const beforeCreation = new Date('2021-06-01T08:51:40Z').valueOf()
    const { kid } = await did.verifyJWS(jws, { atTime: beforeCreation })
    expect(kid).toMatchSnapshot()
  })
})
