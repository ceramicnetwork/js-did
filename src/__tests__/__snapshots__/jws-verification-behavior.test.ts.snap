// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atTime not throw after rotation, if atTime is null 1`] = `
Object {
  "didResolutionResult": Object {
    "didDocument": Object {
      "authentication": Array [
        Object {
          "controller": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl",
          "id": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl#dkbnB3Dd1z5wByn",
          "publicKeyBase58": "fuNBAnTyH62WECeXuDbgnEck7SQZy5i4rFHRKgJxii84",
          "type": "EcdsaSecp256k1Signature2019",
        },
      ],
      "id": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl",
      "keyAgreement": Array [
        Object {
          "controller": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl",
          "id": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl#9t28iZFbczNR2xZ",
          "publicKeyBase58": "thXhSaHPhtJNZaG2pJBYrQ8VChNzYhsFjqa7ALqhfBo",
          "type": "X25519KeyAgreementKey2019",
        },
      ],
      "verificationMethod": Array [
        Object {
          "controller": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl",
          "id": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl#9t28iZFbczNR2xZ",
          "publicKeyBase58": "thXhSaHPhtJNZaG2pJBYrQ8VChNzYhsFjqa7ALqhfBo",
          "type": "X25519KeyAgreementKey2019",
        },
        Object {
          "controller": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl",
          "id": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl#dkbnB3Dd1z5wByn",
          "publicKeyBase58": "fuNBAnTyH62WECeXuDbgnEck7SQZy5i4rFHRKgJxii84",
          "type": "EcdsaSecp256k1Signature2019",
        },
      ],
    },
    "didDocumentMetadata": Object {
      "created": "2021-06-21T08:51:40Z",
      "nextUpdate": "2021-06-21T08:51:40Z",
      "nextVersionId": "bafyreihoahs3ge57ud5kmvs76k2dhuu46fhpilo4r7lol6fuaxh7sxk5nm",
      "versionId": "0",
    },
    "didResolutionMetadata": Object {
      "contentType": "application/did+json",
    },
  },
  "kid": "did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl?version-id=0#dkbnB3Dd1z5wByn",
  "payload": Object {
    "hello": "world",
  },
}
`;

exports[`atTime ok before rotation 1`] = `"did:3:kjzl6cwe1jw14acsxto0wvxwl62usq32y4ncfonzdmg5i5jorjcs2haniuabvzl?version-id=0#dkbnB3Dd1z5wByn"`;
